{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "655ff6bd",
   "metadata": {},
   "source": [
    "# PRS Prep for UKB"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "8318d124",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Imports here.\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import os\n",
    "#import statsmodels.api as sm\n",
    "#import statsmodels.formula.api as smf\n",
    "import warnings\n",
    "warnings.filterwarnings(\"ignore\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "fe161dc8-0fbc-4889-8a2a-da80506e2cb5",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Directory in Biowulf that contains the UKB genetic data\n",
    "os.chdir('/path/to/UKB/data/bfile')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b09e0d19-7b8a-45a9-b350-2d4a6f08cf68",
   "metadata": {},
   "source": [
    "## Activity scores -- overall activity and sedentary traits included\n",
    "https://www.nature.com/articles/s41467-018-07743-4/tables/1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "23f72ed7-35d1-42c7-95ce-c399f3106779",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Status</th>\n",
       "      <th>Trait</th>\n",
       "      <th>ID</th>\n",
       "      <th>Ch</th>\n",
       "      <th>SNP</th>\n",
       "      <th>Position</th>\n",
       "      <th>Nearest gene</th>\n",
       "      <th>Allele (effect/other)</th>\n",
       "      <th>EAF</th>\n",
       "      <th>Beta</th>\n",
       "      <th>SE</th>\n",
       "      <th>p</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Novel</td>\n",
       "      <td>Overall activity</td>\n",
       "      <td>1</td>\n",
       "      <td>10</td>\n",
       "      <td>rs564819152</td>\n",
       "      <td>21,820,650</td>\n",
       "      <td>SKIDA1</td>\n",
       "      <td>A/G</td>\n",
       "      <td>0.679</td>\n",
       "      <td>0.028</td>\n",
       "      <td>0.005</td>\n",
       "      <td>4.200000e-09</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Novel</td>\n",
       "      <td>Sedentary</td>\n",
       "      <td>4</td>\n",
       "      <td>5</td>\n",
       "      <td>rs26579</td>\n",
       "      <td>87,985,295</td>\n",
       "      <td>MEF2C-AS2</td>\n",
       "      <td>G/C</td>\n",
       "      <td>0.415</td>\n",
       "      <td>0.028</td>\n",
       "      <td>0.005</td>\n",
       "      <td>2.600000e-09</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Novel</td>\n",
       "      <td>Sedentary</td>\n",
       "      <td>5</td>\n",
       "      <td>5</td>\n",
       "      <td>rs25981</td>\n",
       "      <td>106,822,908</td>\n",
       "      <td>EFNA5</td>\n",
       "      <td>G/C</td>\n",
       "      <td>0.531</td>\n",
       "      <td>0.028</td>\n",
       "      <td>0.005</td>\n",
       "      <td>3.000000e-09</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Novel</td>\n",
       "      <td>Sedentary</td>\n",
       "      <td>6</td>\n",
       "      <td>3</td>\n",
       "      <td>rs1858242</td>\n",
       "      <td>68,527,135</td>\n",
       "      <td>LOC105377146</td>\n",
       "      <td>A/G</td>\n",
       "      <td>0.259</td>\n",
       "      <td>0.031</td>\n",
       "      <td>0.005</td>\n",
       "      <td>3.100000e-09</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>Novel</td>\n",
       "      <td>Sedentary</td>\n",
       "      <td>7</td>\n",
       "      <td>7</td>\n",
       "      <td>rs34858520</td>\n",
       "      <td>71,723,883</td>\n",
       "      <td>CALN1</td>\n",
       "      <td>A/G</td>\n",
       "      <td>0.558</td>\n",
       "      <td>0.028</td>\n",
       "      <td>0.005</td>\n",
       "      <td>4.200000e-09</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>Known8</td>\n",
       "      <td>Overall activity</td>\n",
       "      <td>8</td>\n",
       "      <td>17</td>\n",
       "      <td>rs2696625</td>\n",
       "      <td>44,326,864</td>\n",
       "      <td>KANSL1-AS1</td>\n",
       "      <td>A/G</td>\n",
       "      <td>0.770</td>\n",
       "      <td>-0.037</td>\n",
       "      <td>0.005</td>\n",
       "      <td>3.200000e-12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>Known8</td>\n",
       "      <td>Overall activity</td>\n",
       "      <td>9</td>\n",
       "      <td>18</td>\n",
       "      <td>rs59499656</td>\n",
       "      <td>40,768,309</td>\n",
       "      <td>SYT4</td>\n",
       "      <td>A/T</td>\n",
       "      <td>0.655</td>\n",
       "      <td>-0.028</td>\n",
       "      <td>0.005</td>\n",
       "      <td>1.900000e-09</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Status             Trait ID  Ch          SNP     Position  Nearest gene  \\\n",
       "0   Novel  Overall activity  1  10  rs564819152   21,820,650        SKIDA1   \n",
       "3   Novel         Sedentary  4   5      rs26579   87,985,295     MEF2C-AS2   \n",
       "4   Novel         Sedentary  5   5      rs25981  106,822,908         EFNA5   \n",
       "5   Novel         Sedentary  6   3    rs1858242   68,527,135  LOC105377146   \n",
       "6   Novel         Sedentary  7   7   rs34858520   71,723,883         CALN1   \n",
       "7  Known8  Overall activity  8  17    rs2696625   44,326,864    KANSL1-AS1   \n",
       "8  Known8  Overall activity  9  18   rs59499656   40,768,309          SYT4   \n",
       "\n",
       "  Allele (effect/other)    EAF   Beta     SE             p  \n",
       "0                   A/G  0.679  0.028  0.005  4.200000e-09  \n",
       "3                   G/C  0.415  0.028  0.005  2.600000e-09  \n",
       "4                   G/C  0.531  0.028  0.005  3.000000e-09  \n",
       "5                   A/G  0.259  0.031  0.005  3.100000e-09  \n",
       "6                   A/G  0.558  0.028  0.005  4.200000e-09  \n",
       "7                   A/G  0.770 -0.037  0.005  3.200000e-12  \n",
       "8                   A/T  0.655 -0.028  0.005  1.900000e-09  "
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#Load table from study - include only \"overall activity\" and \"sedendary\"\n",
    "df = pd.read_csv('/path/to/activity.csv')\n",
    "df = df[df['Trait']!= 'Sleep duration']\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "f6ddb102-a487-4eb2-bdf4-27dab91f2221",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SNP</th>\n",
       "      <th>Allele (effect/other)</th>\n",
       "      <th>Beta</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>rs564819152</td>\n",
       "      <td>A</td>\n",
       "      <td>0.028</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>rs26579</td>\n",
       "      <td>G</td>\n",
       "      <td>0.028</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>rs25981</td>\n",
       "      <td>G</td>\n",
       "      <td>0.028</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>rs1858242</td>\n",
       "      <td>A</td>\n",
       "      <td>0.031</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>rs34858520</td>\n",
       "      <td>A</td>\n",
       "      <td>0.028</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>rs2696625</td>\n",
       "      <td>A</td>\n",
       "      <td>-0.037</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>rs59499656</td>\n",
       "      <td>A</td>\n",
       "      <td>-0.028</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           SNP Allele (effect/other)   Beta\n",
       "0  rs564819152                     A  0.028\n",
       "1      rs26579                     G  0.028\n",
       "2      rs25981                     G  0.028\n",
       "3    rs1858242                     A  0.031\n",
       "4   rs34858520                     A  0.028\n",
       "5    rs2696625                     A -0.037\n",
       "6   rs59499656                     A -0.028"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#Create an \"activity\" file that includes the info needed (\"SNP\", 'Allele (effect)', 'Beta')\n",
    "activity_file = pd.read_csv('/path/to/activity_file.txt', sep = ' ')\n",
    "activity_file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "dcc3d5b2-56a8-44a3-9eec-499c3acb5dee",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{3, 5, 7, 10, 17, 18}"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#Create a list of the chromosomes\n",
    "chrom_list = list(df['Ch'])\n",
    "chrom_list = set(chrom_list)\n",
    "chrom_list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "2c29c1ee-0e45-4a0f-819a-18eb77c55783",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "rs564819152\n",
      "rs26579\n",
      "rs25981\n",
      "rs1858242\n",
      "rs34858520\n",
      "rs2696625\n",
      "rs59499656"
     ]
    }
   ],
   "source": [
    "#Create list of snps; save as activity_snp_list.txt\n",
    "!cat /path/to/activity_snp_list.txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "a3e8763e-8ad0-4da3-906f-cbbac6829b57",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "PLINK v1.90b3.36 64-bit (31 Mar 2016)      https://www.cog-genomics.org/plink2\n",
      "(C) 2005-2016 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to activity_chr3.log.\n",
      "Options in effect:\n",
      "  --bfile /data/CARD/UKBIOBANK/FILTER_IMPUTED_DATA/bfile/chr3.UKBB.EU.filtered\n",
      "  --extract /data/levineks/Virus/activity_snp_list.txt\n",
      "  --make-bed\n",
      "  --out activity_chr3\n",
      "\n",
      "257652 MB RAM detected; reserving 128826 MB for main workspace.\n",
      "649369 variants loaded from .bim file.\n",
      "408961 people (187837 males, 221124 females) loaded from .fam.\n",
      "--extract: 1 variant remaining.\n",
      "Using 1 thread (no multithreaded calculations invoked).\n",
      "Before main variant filters, 408961 founders and 0 nonfounders present.\n",
      "Calculating allele frequencies... 10111213141516171819202122232425262728293031323334353637383940414243444546474849505152535455565758596061626364656667686970717273747576777879808182838485868788899091929394959697989 done.\n",
      "Total genotyping rate is 0.992721.\n",
      "1 variant and 408961 people pass filters and QC.\n",
      "Note: No phenotypes present.\n",
      "--make-bed to activity_chr3.bed + activity_chr3.bim + activity_chr3.fam ...\n",
      "101112131415161718192021222324252627282930313233343536373839404142434445464748495051525354555657585960616263646566676869707172737475767778798081828384858687888990919293949596979899done.\n",
      "PLINK v1.90b3.36 64-bit (31 Mar 2016)      https://www.cog-genomics.org/plink2\n",
      "(C) 2005-2016 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to activity_chr5.log.\n",
      "Options in effect:\n",
      "  --bfile /data/CARD/UKBIOBANK/FILTER_IMPUTED_DATA/bfile/chr5.UKBB.EU.filtered\n",
      "  --extract /data/levineks/Virus/activity_snp_list.txt\n",
      "  --make-bed\n",
      "  --out activity_chr5\n",
      "\n",
      "257652 MB RAM detected; reserving 128826 MB for main workspace.\n",
      "593068 variants loaded from .bim file.\n",
      "408961 people (187837 males, 221124 females) loaded from .fam.\n",
      "--extract: 2 variants remaining.\n",
      "Using 1 thread (no multithreaded calculations invoked).\n",
      "Before main variant filters, 408961 founders and 0 nonfounders present.\n",
      "Calculating allele frequencies... 10111213141516171819202122232425262728293031323334353637383940414243444546474849505152535455565758596061626364656667686970717273747576777879808182838485868788899091929394959697989 done.\n",
      "Total genotyping rate is 0.972194.\n",
      "2 variants and 408961 people pass filters and QC.\n",
      "Note: No phenotypes present.\n",
      "--make-bed to activity_chr5.bed + activity_chr5.bim + activity_chr5.fam ...\n",
      "101112131415161718192021222324252627282930313233343536373839404142434445464748495051525354555657585960616263646566676869707172737475767778798081828384858687888990919293949596979899done.\n",
      "PLINK v1.90b3.36 64-bit (31 Mar 2016)      https://www.cog-genomics.org/plink2\n",
      "(C) 2005-2016 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to activity_chr7.log.\n",
      "Options in effect:\n",
      "  --bfile /data/CARD/UKBIOBANK/FILTER_IMPUTED_DATA/bfile/chr7.UKBB.EU.filtered\n",
      "  --extract /data/levineks/Virus/activity_snp_list.txt\n",
      "  --make-bed\n",
      "  --out activity_chr7\n",
      "\n",
      "257652 MB RAM detected; reserving 128826 MB for main workspace.\n",
      "536346 variants loaded from .bim file.\n",
      "408961 people (187837 males, 221124 females) loaded from .fam.\n",
      "--extract: 1 variant remaining.\n",
      "Using 1 thread (no multithreaded calculations invoked).\n",
      "Before main variant filters, 408961 founders and 0 nonfounders present.\n",
      "Calculating allele frequencies... 10111213141516171819202122232425262728293031323334353637383940414243444546474849505152535455565758596061626364656667686970717273747576777879808182838485868788899091929394959697989 done.\n",
      "Total genotyping rate is 0.990028.\n",
      "1 variant and 408961 people pass filters and QC.\n",
      "Note: No phenotypes present.\n",
      "--make-bed to activity_chr7.bed + activity_chr7.bim + activity_chr7.fam ...\n",
      "101112131415161718192021222324252627282930313233343536373839404142434445464748495051525354555657585960616263646566676869707172737475767778798081828384858687888990919293949596979899done.\n",
      "PLINK v1.90b3.36 64-bit (31 Mar 2016)      https://www.cog-genomics.org/plink2\n",
      "(C) 2005-2016 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to activity_chr10.log.\n",
      "Options in effect:\n",
      "  --bfile /data/CARD/UKBIOBANK/FILTER_IMPUTED_DATA/bfile/chr10.UKBB.EU.filtered\n",
      "  --extract /data/levineks/Virus/activity_snp_list.txt\n",
      "  --make-bed\n",
      "  --out activity_chr10\n",
      "\n",
      "257652 MB RAM detected; reserving 128826 MB for main workspace.\n",
      "461879 variants loaded from .bim file.\n",
      "408961 people (187837 males, 221124 females) loaded from .fam.\n",
      "--extract: 1 variant remaining.\n",
      "Using 1 thread (no multithreaded calculations invoked).\n",
      "Before main variant filters, 408961 founders and 0 nonfounders present.\n",
      "Calculating allele frequencies... 10111213141516171819202122232425262728293031323334353637383940414243444546474849505152535455565758596061626364656667686970717273747576777879808182838485868788899091929394959697989 done.\n",
      "Total genotyping rate is 0.992537.\n",
      "1 variant and 408961 people pass filters and QC.\n",
      "Note: No phenotypes present.\n",
      "--make-bed to activity_chr10.bed + activity_chr10.bim + activity_chr10.fam ...\n",
      "101112131415161718192021222324252627282930313233343536373839404142434445464748495051525354555657585960616263646566676869707172737475767778798081828384858687888990919293949596979899done.\n",
      "PLINK v1.90b3.36 64-bit (31 Mar 2016)      https://www.cog-genomics.org/plink2\n",
      "(C) 2005-2016 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to activity_chr17.log.\n",
      "Options in effect:\n",
      "  --bfile /data/CARD/UKBIOBANK/FILTER_IMPUTED_DATA/bfile/chr17.UKBB.EU.filtered\n",
      "  --extract /data/levineks/Virus/activity_snp_list.txt\n",
      "  --make-bed\n",
      "  --out activity_chr17\n",
      "\n",
      "257652 MB RAM detected; reserving 128826 MB for main workspace.\n",
      "241320 variants loaded from .bim file.\n",
      "408961 people (187837 males, 221124 females) loaded from .fam.\n",
      "--extract: 1 variant remaining.\n",
      "Using 1 thread (no multithreaded calculations invoked).\n",
      "Before main variant filters, 408961 founders and 0 nonfounders present.\n",
      "Calculating allele frequencies... 10111213141516171819202122232425262728293031323334353637383940414243444546474849505152535455565758596061626364656667686970717273747576777879808182838485868788899091929394959697989 done.\n",
      "Total genotyping rate is 0.976788.\n",
      "1 variant and 408961 people pass filters and QC.\n",
      "Note: No phenotypes present.\n",
      "--make-bed to activity_chr17.bed + activity_chr17.bim + activity_chr17.fam ...\n",
      "101112131415161718192021222324252627282930313233343536373839404142434445464748495051525354555657585960616263646566676869707172737475767778798081828384858687888990919293949596979899done.\n",
      "PLINK v1.90b3.36 64-bit (31 Mar 2016)      https://www.cog-genomics.org/plink2\n",
      "(C) 2005-2016 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to activity_chr18.log.\n",
      "Options in effect:\n",
      "  --bfile /data/CARD/UKBIOBANK/FILTER_IMPUTED_DATA/bfile/chr18.UKBB.EU.filtered\n",
      "  --extract /data/levineks/Virus/activity_snp_list.txt\n",
      "  --make-bed\n",
      "  --out activity_chr18\n",
      "\n",
      "257652 MB RAM detected; reserving 128826 MB for main workspace.\n",
      "260248 variants loaded from .bim file.\n",
      "408961 people (187837 males, 221124 females) loaded from .fam.\n",
      "--extract: 1 variant remaining.\n",
      "Using 1 thread (no multithreaded calculations invoked).\n",
      "Before main variant filters, 408961 founders and 0 nonfounders present.\n",
      "Calculating allele frequencies... 10111213141516171819202122232425262728293031323334353637383940414243444546474849505152535455565758596061626364656667686970717273747576777879808182838485868788899091929394959697989 done.\n",
      "Total genotyping rate is 0.991011.\n",
      "1 variant and 408961 people pass filters and QC.\n",
      "Note: No phenotypes present.\n",
      "--make-bed to activity_chr18.bed + activity_chr18.bim + activity_chr18.fam ...\n",
      "101112131415161718192021222324252627282930313233343536373839404142434445464748495051525354555657585960616263646566676869707172737475767778798081828384858687888990919293949596979899done.\n"
     ]
    }
   ],
   "source": [
    "#Extract SNPs for activity\n",
    "activity_chroms = ['3', '5', '7', '10', '17', '18']\n",
    "for i in activity_chroms:\n",
    "    !plink --bfile /path/to/UKBIOBANK/FILTER_IMPUTED_DATA/bfile/chr{i}.UKBB.EU.filtered --extract /data/levineks/Virus/activity_snp_list.txt --make-bed --out activity_chr{i}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "e98e3695-f5b9-4fe4-977e-3956248f3cb0",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/path/to/UKBIOBANK/FILTER_IMPUTED_DATA/bfile/activity_chr3\n",
      "/path/to/UKBIOBANK/FILTER_IMPUTED_DATA/bfile/activity_chr5\n",
      "/path/to/UKBIOBANK/FILTER_IMPUTED_DATA/bfile/activity_chr7\n",
      "/path/to/UKBIOBANK/FILTER_IMPUTED_DATA/bfile/activity_chr10\n",
      "/path/to/UKBIOBANK/FILTER_IMPUTED_DATA/bfile/activity_chr17\n",
      "/path/to/UKBIOBANK/FILTER_IMPUTED_DATA/bfile/activity_chr18\n"
     ]
    }
   ],
   "source": [
    "#Create list of files for merge; save as activity_merge.txt\n",
    "activity_chroms = ['3', '5', '7', '10', '17', '18']\n",
    "for i in activity_chroms:\n",
    "    print(f'/path/to/UKBIOBANK/FILTER_IMPUTED_DATA/bfile/activity_chr{i}')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "b4c35a72-ea42-47a2-8f96-cd37260b5c16",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "PLINK v1.90b3.36 64-bit (31 Mar 2016)      https://www.cog-genomics.org/plink2\n",
      "(C) 2005-2016 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to activity_merged.log.\n",
      "Options in effect:\n",
      "  --make-bed\n",
      "  --merge-list /data/levineks/Virus/activity_merge.txt\n",
      "  --out activity_merged\n",
      "\n",
      "257652 MB RAM detected; reserving 128826 MB for main workspace.\n",
      "Performing single-pass merge (408961 people, 7 variants).\n",
      "Merged fileset written to activity_merged-merge.bed + activity_merged-merge.bim\n",
      "+ activity_merged-merge.fam .\n",
      "7 variants loaded from .bim file.\n",
      "408961 people (187837 males, 221124 females) loaded from .fam.\n",
      "Using 1 thread (no multithreaded calculations invoked).\n",
      "Before main variant filters, 408961 founders and 0 nonfounders present.\n",
      "Calculating allele frequencies... 10111213141516171819202122232425262728293031323334353637383940414243444546474849505152535455565758596061626364656667686970717273747576777879808182838485868788899091929394959697989 done.\n",
      "Total genotyping rate is 0.983925.\n",
      "7 variants and 408961 people pass filters and QC.\n",
      "Note: No phenotypes present.\n",
      "--make-bed to activity_merged.bed + activity_merged.bim + activity_merged.fam\n",
      "... 101112131415161718192021222324252627282930313233343536373839404142434445464748495051525354555657585960616263646566676869707172737475767778798081828384858687888990919293949596979899done.\n"
     ]
    }
   ],
   "source": [
    "#Merge the individual chromosome files\n",
    "!plink --merge-list /data/levineks/Virus/activity_merge.txt --make-bed --out activity_merged"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "9dbe3fe7-3254-4432-910f-6eec906dcd61",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "PLINK v1.90b3.36 64-bit (31 Mar 2016)      https://www.cog-genomics.org/plink2\n",
      "(C) 2005-2016 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to /data/levineks/Virus/activity_prs_test.log.\n",
      "Options in effect:\n",
      "  --bfile /data/CARD/UKBIOBANK/FILTER_IMPUTED_DATA/bfile/activity_merged\n",
      "  --out /data/levineks/Virus/activity_prs_test\n",
      "  --score /data/levineks/Virus/activity_file.txt header\n",
      "\n",
      "257652 MB RAM detected; reserving 128826 MB for main workspace.\n",
      "7 variants loaded from .bim file.\n",
      "408961 people (187837 males, 221124 females) loaded from .fam.\n",
      "Using 1 thread (no multithreaded calculations invoked).\n",
      "Before main variant filters, 408961 founders and 0 nonfounders present.\n",
      "Calculating allele frequencies... 10111213141516171819202122232425262728293031323334353637383940414243444546474849505152535455565758596061626364656667686970717273747576777879808182838485868788899091929394959697989 done.\n",
      "Total genotyping rate is 0.983925.\n",
      "7 variants and 408961 people pass filters and QC.\n",
      "Note: No phenotypes present.\n",
      "--score: 7 valid predictors loaded.\n",
      "--score: Results written to /data/levineks/Virus/activity_prs_test.profile .\n"
     ]
    }
   ],
   "source": [
    "#Get the --score for the merged files\n",
    "!plink --bfile /path/to/UKBIOBANK/FILTER_IMPUTED_DATA/bfile/activity_merged --score /path/to/activity_file.txt header --out /path/to/activity_prs_test"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "637f6834-207f-4159-9d1b-128fe801cb0c",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>FID</th>\n",
       "      <th>IID</th>\n",
       "      <th>PHENO</th>\n",
       "      <th>CNT</th>\n",
       "      <th>CNT2</th>\n",
       "      <th>SCORE</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1000012</td>\n",
       "      <td>1000012</td>\n",
       "      <td>-9</td>\n",
       "      <td>14</td>\n",
       "      <td>7</td>\n",
       "      <td>-0.003071</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1000031</td>\n",
       "      <td>1000031</td>\n",
       "      <td>-9</td>\n",
       "      <td>12</td>\n",
       "      <td>9</td>\n",
       "      <td>0.003053</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1000047</td>\n",
       "      <td>1000047</td>\n",
       "      <td>-9</td>\n",
       "      <td>14</td>\n",
       "      <td>6</td>\n",
       "      <td>0.008000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1000050</td>\n",
       "      <td>1000050</td>\n",
       "      <td>-9</td>\n",
       "      <td>12</td>\n",
       "      <td>7</td>\n",
       "      <td>0.007204</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1000068</td>\n",
       "      <td>1000068</td>\n",
       "      <td>-9</td>\n",
       "      <td>14</td>\n",
       "      <td>8</td>\n",
       "      <td>-0.001286</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>408956</th>\n",
       "      <td>6026129</td>\n",
       "      <td>6026129</td>\n",
       "      <td>-9</td>\n",
       "      <td>14</td>\n",
       "      <td>8</td>\n",
       "      <td>-0.001286</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>408957</th>\n",
       "      <td>6026131</td>\n",
       "      <td>6026131</td>\n",
       "      <td>-9</td>\n",
       "      <td>14</td>\n",
       "      <td>8</td>\n",
       "      <td>0.003357</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>408958</th>\n",
       "      <td>6026147</td>\n",
       "      <td>6026147</td>\n",
       "      <td>-9</td>\n",
       "      <td>14</td>\n",
       "      <td>8</td>\n",
       "      <td>0.007786</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>408959</th>\n",
       "      <td>6026150</td>\n",
       "      <td>6026150</td>\n",
       "      <td>-9</td>\n",
       "      <td>14</td>\n",
       "      <td>8</td>\n",
       "      <td>0.002714</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>408960</th>\n",
       "      <td>6026168</td>\n",
       "      <td>6026168</td>\n",
       "      <td>-9</td>\n",
       "      <td>14</td>\n",
       "      <td>7</td>\n",
       "      <td>0.000714</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>408961 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            FID      IID  PHENO  CNT  CNT2     SCORE\n",
       "0       1000012  1000012     -9   14     7 -0.003071\n",
       "1       1000031  1000031     -9   12     9  0.003053\n",
       "2       1000047  1000047     -9   14     6  0.008000\n",
       "3       1000050  1000050     -9   12     7  0.007204\n",
       "4       1000068  1000068     -9   14     8 -0.001286\n",
       "...         ...      ...    ...  ...   ...       ...\n",
       "408956  6026129  6026129     -9   14     8 -0.001286\n",
       "408957  6026131  6026131     -9   14     8  0.003357\n",
       "408958  6026147  6026147     -9   14     8  0.007786\n",
       "408959  6026150  6026150     -9   14     8  0.002714\n",
       "408960  6026168  6026168     -9   14     7  0.000714\n",
       "\n",
       "[408961 rows x 6 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#Take a look at the results\n",
    "activity = pd.read_csv('/path/to/activity_prs_test.profile', sep = '\\s+')\n",
    "activity"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d0176e11-18e9-416a-95f9-b14a4fbb8cfb",
   "metadata": {},
   "source": [
    "## Diet scores -- protein included\n",
    "https://www.ncbi.nlm.nih.gov/pmc/articles/PMC7767645/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "5d3abe67-c81a-472e-8172-e760e4d3bd52",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Top hit in locus for</th>\n",
       "      <th>SNPID</th>\n",
       "      <th>CHR</th>\n",
       "      <th>BP</th>\n",
       "      <th>Effect allele</th>\n",
       "      <th>Beta</th>\n",
       "      <th>P-value</th>\n",
       "      <th>Nearest gene</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Protein</td>\n",
       "      <td>rs780094</td>\n",
       "      <td>2.0</td>\n",
       "      <td>27,741,237</td>\n",
       "      <td>t</td>\n",
       "      <td>0.018</td>\n",
       "      <td>5.580000e-10</td>\n",
       "      <td>GCKR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Protein</td>\n",
       "      <td>rs445551</td>\n",
       "      <td>2.0</td>\n",
       "      <td>79,697,982</td>\n",
       "      <td>a</td>\n",
       "      <td>0.019</td>\n",
       "      <td>1.490000e-08</td>\n",
       "      <td>CTNNA2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Protein</td>\n",
       "      <td>rs1603978</td>\n",
       "      <td>3.0</td>\n",
       "      <td>25,108,236</td>\n",
       "      <td>a</td>\n",
       "      <td>0.019</td>\n",
       "      <td>1.350000e-10</td>\n",
       "      <td>AC092422.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>Protein</td>\n",
       "      <td>rs13146907</td>\n",
       "      <td>4.0</td>\n",
       "      <td>39,425,248</td>\n",
       "      <td>a</td>\n",
       "      <td>−0.022</td>\n",
       "      <td>1.240000e-14</td>\n",
       "      <td>KLB</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>Protein</td>\n",
       "      <td>rs1461729</td>\n",
       "      <td>8.0</td>\n",
       "      <td>9,187,242</td>\n",
       "      <td>a</td>\n",
       "      <td>0.032</td>\n",
       "      <td>4.090000e-12</td>\n",
       "      <td>AC022784.6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>Protein</td>\n",
       "      <td>rs55872725</td>\n",
       "      <td>16.0</td>\n",
       "      <td>53,809,123</td>\n",
       "      <td>t</td>\n",
       "      <td>0.018</td>\n",
       "      <td>2.090000e-10</td>\n",
       "      <td>FTO</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>Protein</td>\n",
       "      <td>rs838133</td>\n",
       "      <td>19.0</td>\n",
       "      <td>49,259,529</td>\n",
       "      <td>a</td>\n",
       "      <td>−0.032</td>\n",
       "      <td>4.520000e-26</td>\n",
       "      <td>FGF21</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Top hit in locus for       SNPID   CHR          BP Effect allele    Beta  \\\n",
       "0               Protein    rs780094   2.0  27,741,237             t   0.018   \n",
       "3               Protein    rs445551   2.0  79,697,982             a   0.019   \n",
       "5               Protein   rs1603978   3.0  25,108,236             a   0.019   \n",
       "8               Protein  rs13146907   4.0  39,425,248             a  −0.022   \n",
       "16              Protein   rs1461729   8.0   9,187,242             a   0.032   \n",
       "19              Protein  rs55872725  16.0  53,809,123             t   0.018   \n",
       "34              Protein    rs838133  19.0  49,259,529             a  −0.032   \n",
       "\n",
       "         P-value Nearest gene  \n",
       "0   5.580000e-10         GCKR  \n",
       "3   1.490000e-08       CTNNA2  \n",
       "5   1.350000e-10   AC092422.1  \n",
       "8   1.240000e-14          KLB  \n",
       "16  4.090000e-12   AC022784.6  \n",
       "19  2.090000e-10          FTO  \n",
       "34  4.520000e-26        FGF21  "
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#Load table from study - include only \"protein\"\n",
    "df = pd.read_csv('/path/to/diet.csv')\n",
    "df = df[df['Top hit in locus for']=='Protein']\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "92588a1c-eb02-45dd-beb9-948e2deddd3d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SNPID</th>\n",
       "      <th>Effect allele</th>\n",
       "      <th>Beta</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>rs780094</td>\n",
       "      <td>T</td>\n",
       "      <td>0.018</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>rs445551</td>\n",
       "      <td>A</td>\n",
       "      <td>0.019</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>rs1603978</td>\n",
       "      <td>A</td>\n",
       "      <td>0.019</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>rs13146907</td>\n",
       "      <td>A</td>\n",
       "      <td>−0.022</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>rs1461729</td>\n",
       "      <td>A</td>\n",
       "      <td>0.032</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>rs55872725</td>\n",
       "      <td>T</td>\n",
       "      <td>0.018</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>rs838133</td>\n",
       "      <td>A</td>\n",
       "      <td>−0.032</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        SNPID Effect allele    Beta\n",
       "0    rs780094             T   0.018\n",
       "1    rs445551             A   0.019\n",
       "2   rs1603978             A   0.019\n",
       "3  rs13146907             A  −0.022\n",
       "4   rs1461729             A   0.032\n",
       "5  rs55872725             T   0.018\n",
       "6    rs838133             A  −0.032"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#Pick the correct columns and save to file - 'SNPID', 'Effect allele', 'Beta'\n",
    "#Remember to make effect alleles UPPER CASE\n",
    "diet_file = pd.read_csv('/path/to/diet_file.txt', sep = ' ')\n",
    "diet_file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "891a02c4-77bb-473b-b629-5fed8d47d4bb",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{2.0, 3.0, 4.0, 8.0, 16.0, 19.0}"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#Create a list of the chromosomes\n",
    "chrom_list = list(df['CHR'])\n",
    "chrom_list = set(chrom_list)\n",
    "chrom_list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "ab468b6c-2016-44d4-8d3d-8005e1110a05",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "rs780094\n",
      "rs445551\n",
      "rs1603978\n",
      "rs13146907\n",
      "rs1461729\n",
      "rs55872725\n",
      "rs838133"
     ]
    }
   ],
   "source": [
    "#Create a list of SNPs; save to diet_snp_list.txt\n",
    "!cat /path/to/Virus/diet_snp_list.txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "da6f9a62-1fe5-4670-b9ea-738fc9095228",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "PLINK v1.90b3.36 64-bit (31 Mar 2016)      https://www.cog-genomics.org/plink2\n",
      "(C) 2005-2016 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to diet_chr2.log.\n",
      "Options in effect:\n",
      "  --bfile /data/CARD/UKBIOBANK/FILTER_IMPUTED_DATA/bfile/chr2.UKBB.EU.filtered\n",
      "  --extract /data/levineks/Virus/diet_snp_list.txt\n",
      "  --make-bed\n",
      "  --out diet_chr2\n",
      "\n",
      "257652 MB RAM detected; reserving 128826 MB for main workspace.\n",
      "764315 variants loaded from .bim file.\n",
      "408961 people (187837 males, 221124 females) loaded from .fam.\n",
      "--extract: 2 variants remaining.\n",
      "Using 1 thread (no multithreaded calculations invoked).\n",
      "Before main variant filters, 408961 founders and 0 nonfounders present.\n",
      "Calculating allele frequencies... 10111213141516171819202122232425262728293031323334353637383940414243444546474849505152535455565758596061626364656667686970717273747576777879808182838485868788899091929394959697989 done.\n",
      "2 variants and 408961 people pass filters and QC.\n",
      "Note: No phenotypes present.\n",
      "--make-bed to diet_chr2.bed + diet_chr2.bim + diet_chr2.fam ... 101112131415161718192021222324252627282930313233343536373839404142434445464748495051525354555657585960616263646566676869707172737475767778798081828384858687888990919293949596979899done.\n",
      "PLINK v1.90b3.36 64-bit (31 Mar 2016)      https://www.cog-genomics.org/plink2\n",
      "(C) 2005-2016 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to diet_chr3.log.\n",
      "Options in effect:\n",
      "  --bfile /data/CARD/UKBIOBANK/FILTER_IMPUTED_DATA/bfile/chr3.UKBB.EU.filtered\n",
      "  --extract /data/levineks/Virus/diet_snp_list.txt\n",
      "  --make-bed\n",
      "  --out diet_chr3\n",
      "\n",
      "257652 MB RAM detected; reserving 128826 MB for main workspace.\n",
      "649369 variants loaded from .bim file.\n",
      "408961 people (187837 males, 221124 females) loaded from .fam.\n",
      "--extract: 1 variant remaining.\n",
      "Using 1 thread (no multithreaded calculations invoked).\n",
      "Before main variant filters, 408961 founders and 0 nonfounders present.\n",
      "Calculating allele frequencies... 10111213141516171819202122232425262728293031323334353637383940414243444546474849505152535455565758596061626364656667686970717273747576777879808182838485868788899091929394959697989 done.\n",
      "Total genotyping rate is 0.988124.\n",
      "1 variant and 408961 people pass filters and QC.\n",
      "Note: No phenotypes present.\n",
      "--make-bed to diet_chr3.bed + diet_chr3.bim + diet_chr3.fam ... 101112131415161718192021222324252627282930313233343536373839404142434445464748495051525354555657585960616263646566676869707172737475767778798081828384858687888990919293949596979899done.\n",
      "PLINK v1.90b3.36 64-bit (31 Mar 2016)      https://www.cog-genomics.org/plink2\n",
      "(C) 2005-2016 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to diet_chr4.log.\n",
      "Options in effect:\n",
      "  --bfile /data/CARD/UKBIOBANK/FILTER_IMPUTED_DATA/bfile/chr4.UKBB.EU.filtered\n",
      "  --extract /data/levineks/Virus/diet_snp_list.txt\n",
      "  --make-bed\n",
      "  --out diet_chr4\n",
      "\n",
      "257652 MB RAM detected; reserving 128826 MB for main workspace.\n",
      "668008 variants loaded from .bim file.\n",
      "408961 people (187837 males, 221124 females) loaded from .fam.\n",
      "--extract: 1 variant remaining.\n",
      "Using 1 thread (no multithreaded calculations invoked).\n",
      "Before main variant filters, 408961 founders and 0 nonfounders present.\n",
      "Calculating allele frequencies... 10111213141516171819202122232425262728293031323334353637383940414243444546474849505152535455565758596061626364656667686970717273747576777879808182838485868788899091929394959697989 done.\n",
      "1 variant and 408961 people pass filters and QC.\n",
      "Note: No phenotypes present.\n",
      "--make-bed to diet_chr4.bed + diet_chr4.bim + diet_chr4.fam ... 101112131415161718192021222324252627282930313233343536373839404142434445464748495051525354555657585960616263646566676869707172737475767778798081828384858687888990919293949596979899done.\n",
      "PLINK v1.90b3.36 64-bit (31 Mar 2016)      https://www.cog-genomics.org/plink2\n",
      "(C) 2005-2016 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to diet_chr8.log.\n",
      "Options in effect:\n",
      "  --bfile /data/CARD/UKBIOBANK/FILTER_IMPUTED_DATA/bfile/chr8.UKBB.EU.filtered\n",
      "  --extract /data/levineks/Virus/diet_snp_list.txt\n",
      "  --make-bed\n",
      "  --out diet_chr8\n",
      "\n",
      "257652 MB RAM detected; reserving 128826 MB for main workspace.\n",
      "502726 variants loaded from .bim file.\n",
      "408961 people (187837 males, 221124 females) loaded from .fam.\n",
      "--extract: 1 variant remaining.\n",
      "Using 1 thread (no multithreaded calculations invoked).\n",
      "Before main variant filters, 408961 founders and 0 nonfounders present.\n",
      "Calculating allele frequencies... 10111213141516171819202122232425262728293031323334353637383940414243444546474849505152535455565758596061626364656667686970717273747576777879808182838485868788899091929394959697989 done.\n",
      "Total genotyping rate is 0.999672.\n",
      "1 variant and 408961 people pass filters and QC.\n",
      "Note: No phenotypes present.\n",
      "--make-bed to diet_chr8.bed + diet_chr8.bim + diet_chr8.fam ... 101112131415161718192021222324252627282930313233343536373839404142434445464748495051525354555657585960616263646566676869707172737475767778798081828384858687888990919293949596979899done.\n",
      "PLINK v1.90b3.36 64-bit (31 Mar 2016)      https://www.cog-genomics.org/plink2\n",
      "(C) 2005-2016 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to diet_chr16.log.\n",
      "Options in effect:\n",
      "  --bfile /data/CARD/UKBIOBANK/FILTER_IMPUTED_DATA/bfile/chr16.UKBB.EU.filtered\n",
      "  --extract /data/levineks/Virus/diet_snp_list.txt\n",
      "  --make-bed\n",
      "  --out diet_chr16\n",
      "\n",
      "257652 MB RAM detected; reserving 128826 MB for main workspace.\n",
      "275744 variants loaded from .bim file.\n",
      "408961 people (187837 males, 221124 females) loaded from .fam.\n",
      "--extract: 1 variant remaining.\n",
      "Using 1 thread (no multithreaded calculations invoked).\n",
      "Before main variant filters, 408961 founders and 0 nonfounders present.\n",
      "Calculating allele frequencies... 10111213141516171819202122232425262728293031323334353637383940414243444546474849505152535455565758596061626364656667686970717273747576777879808182838485868788899091929394959697989 done.\n",
      "Total genotyping rate is 0.999773.\n",
      "1 variant and 408961 people pass filters and QC.\n",
      "Note: No phenotypes present.\n",
      "--make-bed to diet_chr16.bed + diet_chr16.bim + diet_chr16.fam ... 101112131415161718192021222324252627282930313233343536373839404142434445464748495051525354555657585960616263646566676869707172737475767778798081828384858687888990919293949596979899done.\n",
      "PLINK v1.90b3.36 64-bit (31 Mar 2016)      https://www.cog-genomics.org/plink2\n",
      "(C) 2005-2016 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to diet_chr19.log.\n",
      "Options in effect:\n",
      "  --bfile /data/CARD/UKBIOBANK/FILTER_IMPUTED_DATA/bfile/chr19.UKBB.EU.filtered\n",
      "  --extract /data/levineks/Virus/diet_snp_list.txt\n",
      "  --make-bed\n",
      "  --out diet_chr19\n",
      "\n",
      "257652 MB RAM detected; reserving 128826 MB for main workspace.\n",
      "206654 variants loaded from .bim file.\n",
      "408961 people (187837 males, 221124 females) loaded from .fam.\n",
      "Error: No variants remaining after --extract.\n"
     ]
    }
   ],
   "source": [
    "#Extract SNPs for diet\n",
    "diet_chroms = ['2', '3', '4', '8', '16', '19']\n",
    "for i in diet_chroms:\n",
    "    !plink --bfile /path/to/UKBIOBANK/FILTER_IMPUTED_DATA/bfile/chr{i}.UKBB.EU.filtered --extract /path/to/diet_snp_list.txt --make-bed --out diet_chr{i}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "cc66312b-d754-4dbb-a738-2d44fc965091",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/path/to/UKBIOBANK/FILTER_IMPUTED_DATA/bfile/diet_chr2\n",
      "/path/to/UKBIOBANK/FILTER_IMPUTED_DATA/bfile/diet_chr3\n",
      "/path/to/UKBIOBANK/FILTER_IMPUTED_DATA/bfile/diet_chr4\n",
      "/path/to/UKBIOBANK/FILTER_IMPUTED_DATA/bfile/diet_chr8\n",
      "/path/to/UKBIOBANK/FILTER_IMPUTED_DATA/bfile/diet_chr16\n"
     ]
    }
   ],
   "source": [
    "##Error: No variants remaining after --extract for chr19 -- Remove #19\n",
    "#Create list of files for merge; save as diet_merge.txt\n",
    "diet_chroms = ['2', '3', '4', '8', '16']\n",
    "for i in diet_chroms:\n",
    "    print(f'/path/to/UKBIOBANK/FILTER_IMPUTED_DATA/bfile/diet_chr{i}')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "f157edef-f4a8-4828-ae0c-14f0838f28b8",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "PLINK v1.90b3.36 64-bit (31 Mar 2016)      https://www.cog-genomics.org/plink2\n",
      "(C) 2005-2016 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to diet_merged.log.\n",
      "Options in effect:\n",
      "  --make-bed\n",
      "  --merge-list /data/levineks/Virus/diet_merge.txt\n",
      "  --out diet_merged\n",
      "\n",
      "257652 MB RAM detected; reserving 128826 MB for main workspace.\n",
      "Performing single-pass merge (408961 people, 6 variants).\n",
      "Merged fileset written to diet_merged-merge.bed + diet_merged-merge.bim +\n",
      "diet_merged-merge.fam .\n",
      "6 variants loaded from .bim file.\n",
      "408961 people (187837 males, 221124 females) loaded from .fam.\n",
      "Using 1 thread (no multithreaded calculations invoked).\n",
      "Before main variant filters, 408961 founders and 0 nonfounders present.\n",
      "Calculating allele frequencies... 10111213141516171819202122232425262728293031323334353637383940414243444546474849505152535455565758596061626364656667686970717273747576777879808182838485868788899091929394959697989 done.\n",
      "Total genotyping rate is 0.997928.\n",
      "6 variants and 408961 people pass filters and QC.\n",
      "Note: No phenotypes present.\n",
      "--make-bed to diet_merged.bed + diet_merged.bim + diet_merged.fam ... 101112131415161718192021222324252627282930313233343536373839404142434445464748495051525354555657585960616263646566676869707172737475767778798081828384858687888990919293949596979899done.\n"
     ]
    }
   ],
   "source": [
    "#Merge the individual chromosome files\n",
    "!plink --merge-list /path/to/diet_merge.txt --make-bed --out diet_merged"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "936dc774-9e6d-485a-bee3-f9674ba51b23",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "PLINK v1.90b3.36 64-bit (31 Mar 2016)      https://www.cog-genomics.org/plink2\n",
      "(C) 2005-2016 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to /data/levineks/Virus/diet_prs_test.log.\n",
      "Options in effect:\n",
      "  --bfile /data/CARD/UKBIOBANK/FILTER_IMPUTED_DATA/bfile/diet_merged\n",
      "  --out /data/levineks/Virus/diet_prs_test\n",
      "  --score /data/levineks/Virus/diet_file.txt header\n",
      "\n",
      "257652 MB RAM detected; reserving 128826 MB for main workspace.\n",
      "6 variants loaded from .bim file.\n",
      "408961 people (187837 males, 221124 females) loaded from .fam.\n",
      "Using 1 thread (no multithreaded calculations invoked).\n",
      "Before main variant filters, 408961 founders and 0 nonfounders present.\n",
      "Calculating allele frequencies... 10111213141516171819202122232425262728293031323334353637383940414243444546474849505152535455565758596061626364656667686970717273747576777879808182838485868788899091929394959697989 done.\n",
      "Total genotyping rate is 0.997928.\n",
      "6 variants and 408961 people pass filters and QC.\n",
      "Note: No phenotypes present.\n",
      "Warning: 1 line skipped in --score file (1 due to variant ID mismatch, 0 due to\n",
      "allele code mismatch); see /data/levineks/Virus/diet_prs_test.nopred for\n",
      "details.\n",
      "--score: 6 valid predictors loaded.\n",
      "--score: Results written to /data/levineks/Virus/diet_prs_test.profile .\n"
     ]
    }
   ],
   "source": [
    "#Get the --score for the merged files\n",
    "!plink --bfile /path/to/UKBIOBANK/FILTER_IMPUTED_DATA/bfile/diet_merged --score /path/to/diet_file.txt header --out /path/to/diet_prs_test"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "180fc0da-8c50-4306-a3b4-cd4722b9d9f4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>FID</th>\n",
       "      <th>IID</th>\n",
       "      <th>PHENO</th>\n",
       "      <th>CNT</th>\n",
       "      <th>CNT2</th>\n",
       "      <th>SCORE</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1000012</td>\n",
       "      <td>1000012</td>\n",
       "      <td>-9</td>\n",
       "      <td>12</td>\n",
       "      <td>5</td>\n",
       "      <td>0.004250</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1000031</td>\n",
       "      <td>1000031</td>\n",
       "      <td>-9</td>\n",
       "      <td>12</td>\n",
       "      <td>3</td>\n",
       "      <td>0.001250</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1000047</td>\n",
       "      <td>1000047</td>\n",
       "      <td>-9</td>\n",
       "      <td>12</td>\n",
       "      <td>5</td>\n",
       "      <td>0.004333</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1000050</td>\n",
       "      <td>1000050</td>\n",
       "      <td>-9</td>\n",
       "      <td>12</td>\n",
       "      <td>4</td>\n",
       "      <td>0.006250</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1000068</td>\n",
       "      <td>1000068</td>\n",
       "      <td>-9</td>\n",
       "      <td>12</td>\n",
       "      <td>4</td>\n",
       "      <td>0.003917</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>408956</th>\n",
       "      <td>6026129</td>\n",
       "      <td>6026129</td>\n",
       "      <td>-9</td>\n",
       "      <td>12</td>\n",
       "      <td>7</td>\n",
       "      <td>0.007500</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>408957</th>\n",
       "      <td>6026131</td>\n",
       "      <td>6026131</td>\n",
       "      <td>-9</td>\n",
       "      <td>12</td>\n",
       "      <td>4</td>\n",
       "      <td>0.002833</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>408958</th>\n",
       "      <td>6026147</td>\n",
       "      <td>6026147</td>\n",
       "      <td>-9</td>\n",
       "      <td>12</td>\n",
       "      <td>6</td>\n",
       "      <td>0.005917</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>408959</th>\n",
       "      <td>6026150</td>\n",
       "      <td>6026150</td>\n",
       "      <td>-9</td>\n",
       "      <td>12</td>\n",
       "      <td>8</td>\n",
       "      <td>0.006750</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>408960</th>\n",
       "      <td>6026168</td>\n",
       "      <td>6026168</td>\n",
       "      <td>-9</td>\n",
       "      <td>12</td>\n",
       "      <td>4</td>\n",
       "      <td>0.002917</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>408961 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            FID      IID  PHENO  CNT  CNT2     SCORE\n",
       "0       1000012  1000012     -9   12     5  0.004250\n",
       "1       1000031  1000031     -9   12     3  0.001250\n",
       "2       1000047  1000047     -9   12     5  0.004333\n",
       "3       1000050  1000050     -9   12     4  0.006250\n",
       "4       1000068  1000068     -9   12     4  0.003917\n",
       "...         ...      ...    ...  ...   ...       ...\n",
       "408956  6026129  6026129     -9   12     7  0.007500\n",
       "408957  6026131  6026131     -9   12     4  0.002833\n",
       "408958  6026147  6026147     -9   12     6  0.005917\n",
       "408959  6026150  6026150     -9   12     8  0.006750\n",
       "408960  6026168  6026168     -9   12     4  0.002917\n",
       "\n",
       "[408961 rows x 6 columns]"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#Take a look at the results\n",
    "diet = pd.read_csv('/path/to/diet_prs_test.profile', sep = '\\s+')\n",
    "diet"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "282c3eaf-cf84-4d1d-b152-bbbcbd973642",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
